libs/ballot-interpreter/src/hmpb-rust/timing_marks/corners/border_finding.rs

```diff
@@ -0,0 +1,202 @@
+                reason: format!("Left timing mark border has an unexpected number of marks. Expected {} marks, found {}", geometry.grid_size.height, actual_left_count),
+            });
+        }
+
+        let actual_right_count = right.marks.len();
+        if actual_right_count != geometry.grid_size.height as usize {
+            return Err(Error::MissingTimingMarks {
+                reason: format!("Right timing mark border has an unexpected number of marks. Expected {} marks, found {}", geometry.grid_size.height, actual_right_count),
+            });
+        }
+
+        let actual_top_count = top.marks.len();
+        if actual_top_count != geometry.grid_size.width as usize {
+            return Err(Error::MissingTimingMarks {
+                reason: format!("Top timing mark border has an unexpected number of marks. Expected {} marks, found {}", geometry.grid_size.width, actual_top_count),
+            });
+        }
+
+        let actual_bottom_count = bottom.marks.len();
+        if actual_bottom_count != geometry.grid_size.width as usize {
+            return Err(Error::MissingTimingMarks {
+                reason: format!("Bottom timing mark border has an unexpected number of marks. Expected {} marks, found {}", geometry.grid_size.width, actual_bottom_count),
+            });
+        }
+
+        Ok(Self {
+            left,
+            right,
+            top,
+            bottom,
+        })
+    }
+
+    pub fn debug_draw(&self, canvas: &mut RgbImage) {
+        for (mark, color) in self
+            .left
+            .marks
+            .iter()
+            .chain(self.right.marks.iter())
+            .chain(self.top.marks.iter())
+            .chain(self.bottom.marks.iter())
+            .zip(rainbow())
+        {
+            draw_filled_rect_mut(canvas, imageproc_rect_from_rect(&mark.rect()), color);
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GridBorder {
+    border: Border,
+    marks: Vec<CandidateTimingMark>,
+}
+
+impl GridBorder {
+    /// Find the set of marks along a border by moving along the space between
+    /// two corners, finding timing marks that are close enough to the expected
+    /// location to be counted.
+    pub fn find_between_corners(
+        geometry: &Geometry,
+        border: Border,
+        candidate_timing_marks: &[CandidateTimingMark],
+        corners: (&CandidateTimingMark, &CandidateTimingMark),
+    ) -> Result<Self, Error> {
+        assert!(
+            candidate_timing_marks.contains(corners.0),
+            "Starting corner must be a candidate timing mark"
+        );
+        assert!(
+            candidate_timing_marks.contains(corners.1),
+            "Ending corner must be a candidate timing mark"
+        );
+
+        let border = border;
+        let (corner1, corner2) = corners;
+        let timing_mark_center_to_center_distance =
+            geometry.timing_mark_size.width + geometry.timing_mark_size.height;
```

Comment:
Same note about risk of this distance being different across page sizes. Also maybe should be factored since if it's being used in multiple places and is a sensitive part of the algorithm.

Also, is the distance the same for left/right columns and top/bottom rows?

Reply:
As noted in [7ef20a2](https://github.com/votingworks/vxsuite/pull/6582/commits/7ef20a2e5dd0e393a574effcd81b165edf212cf8), the distances may be different horizontally vs vertically.